test/unit/org/apache/cassandra/locator/AssureSufficientLiveNodesTest.java

```diff
@@ -0,0 +1,317 @@
+    private static final int RACE_TEST_LOOPS = 100;
+    private static final Token tk = new Murmur3Partitioner.LongToken(0);
+
+    @BeforeClass
+    public static void setUpClass() throws Throwable
+    {
+        SchemaLoader.loadSchema();
+        // Register peers with expected DC for NetworkTopologyStrategy.
+        TokenMetadata metadata = StorageService.instance.getTokenMetadata();
+        metadata.clearUnsafe();
+
+        DatabaseDescriptor.setEndpointSnitch(new AbstractNetworkTopologySnitch()
+        {
+            public String getRack(InetAddressAndPort endpoint)
+            {
+                byte[] address = endpoint.addressBytes;
+                return "rake" + address[1];
+            }
+
+            public String getDatacenter(InetAddressAndPort endpoint)
+            {
+                byte[] address = endpoint.addressBytes;
+                return "datacenter" + address[1];
+            }
+        });
+
+        List<InetAddressAndPort> instances = ImmutableList.of(
+            // datacenter 1
+            InetAddressAndPort.getByName("127.1.0.255"), InetAddressAndPort.getByName("127.1.0.254"), InetAddressAndPort.getByName("127.1.0.253"),
+            // datacenter 2
+            InetAddressAndPort.getByName("127.2.0.255"), InetAddressAndPort.getByName("127.2.0.254"), InetAddressAndPort.getByName("127.2.0.253"),
+            // datacenter 3
+            InetAddressAndPort.getByName("127.3.0.255"), InetAddressAndPort.getByName("127.3.0.254"), InetAddressAndPort.getByName("127.3.0.253"));
+
+        for (int i = 0; i < instances.size(); i++)
+        {
+            InetAddressAndPort ip = instances.get(i);
+            metadata.updateHostId(UUID.randomUUID(), ip);
+            metadata.updateNormalToken(new Murmur3Partitioner.LongToken(i), ip);
+        }
+    }
+
+    @Test
+    public void insufficientLiveNodesTest()
+    {
+        final KeyspaceParams largeRF = KeyspaceParams.nts("datacenter1", 6);
+        // Not a race in fact. It is just testing the Unavailable can be correctly thrown.
+        assertThatThrownBy(() -> {
+            raceOfReplicationStrategyTest(largeRF, largeRF, 1,
+                                          keyspace -> ReplicaPlans.forWrite(keyspace, QUORUM, tk, ReplicaPlans.writeNormal));
+        }).as("Unavailable should be thrown given 3 live nodes is less than a quorum of 6")
+          .isInstanceOf(UnavailableException.class)
+          .hasMessageContaining("Cannot achieve consistency level QUORUM");
+    }
+
+    @Test
+    public void addDatacenterShouldNotCausesUnavailableWithEachQuorumTest() throws Throwable
+    {
+        // write
+        raceOfReplicationStrategyTest(
+            // init
+            KeyspaceParams.nts(DC1, 3),
+            // alter to
+            KeyspaceParams.nts(DC1, 3, DC2, 3),
+            // test
+            keyspace -> {
+                ReplicaPlans.forWrite(keyspace, EACH_QUORUM, tk, ReplicaPlans.writeNormal);
+            }
```

Comment:
Nit: we don't need the braces
```suggestion
            keyspace -> ReplicaPlans.forWrite(keyspace, EACH_QUORUM, tk, ReplicaPlans.writeNormal)
```